### PR TITLE
Add Rescheduler addon

### LINKED ISSUE #944: guaranteed scheduling of critical add-ons requires a **Rescheduler** pod
---

**Is this an ISSUE or FEATURE REQUEST?** ISSUE
The kube-dns ( and also heapster and calico) deployment config yaml contain the following annotation:
      `annotations:
        scheduler.alpha.kubernetes.io/critical-pod: ""
        scheduler.alpha.kubernetes.io/tolerations: "[{\"key\":\"CriticalAddonsOnly\",\"operator\":\"Exists\"}]"`
According to the docs here: https://kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/ , guaranteed scheduling of critical add-ons requires a **Rescheduler** pod to be running. An example Rescheduler manifest is available here: https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/rescheduler/rescheduler.manifest
 
I am unable to find a rescheduler pod in acs-engine. Is the rescheduler no longer required ?

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
kubernetes 1.6.2

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What this PR does / why we need it**: Guaranteed scheduling of critical add-ons requires a Rescheduler pod to be running, this PR adds that Rescheduler.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #944 

**Release note**:

### CODE REVIEW

File: pkg/acsengine/defaults.go [hunk 1]
```diff
@@ -178,6 +178,21 @@ var (
 			},
 		},
 	}
+
+	// DefaultReschedulerAddonsConfig is the default rescheduler Kubernetes addon Config
+	DefaultReschedulerAddonsConfig = api.KubernetesAddon{
+		Name:    DefaultReschedulerAddonName,
+		Enabled: pointerToBool(true),
```
Comment (CecileRobertMichon) [inline pkg/acsengine/defaults.go]: Should this be enabled by default?

File: pkg/acsengine/defaults.go [hunk 2]
```diff
@@ -178,6 +178,21 @@ var (
...
+		Containers: []api.KubernetesContainerSpec{
+			{
+				Name:           DefaultReschedulerAddonName,
+				CPURequests:    "300m",
+				MemoryRequests: "150Mi",
+				CPULimits:      "300m",
+				MemoryLimits:   "150Mi",
```
Comment (CecileRobertMichon) [inline pkg/acsengine/defaults.go]: ^^ need to discuss what default limits to use for rescheduler

File: parts/kubernetesmasteraddons-kube-rescheduler-deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        k8s-app: rescheduler
+      annotations:
+        scheduler.alpha.kubernetes.io/critical-pod: ''
+    spec:
+      containers:
+      - image: <kubernetesReschedulerSpec>
+        name: rescheduler
+        resources:
+          requests:
+            cpu: 10m
```
Comment (CecileRobertMichon) [inline parts/kubernetesmasteraddons-kube-rescheduler-deployment.yaml]: // todo change this to param

File: pkg/acsengine/engine.go [hunk 4]
```diff
@@ -570,7 +572,20 @@ func getParameters(cs *api.ContainerService, isClassicMode bool, generatorCode s
 			if dashboardAddon.Containers[c].Image != "" {
 				addValue(parametersMap, "kubernetesDashboardSpec", dashboardAddon.Containers[c].Image)
 			} else {
-				addValue(parametersMap, "kubernetesDashboardSpec", cloudSpecConfig.KubernetesSpecConfig.KubernetesImageBase+KubeConfigs[k8sVersion]["dashboard"])
+				addValue(parametersMap, "kubernetesDashboardSpec", cloudSpecConfig.KubernetesSpecConfig.KubernetesImageBase+KubeConfigs[k8sVersion][DefaultDashboardAddonName])
```
Comment (CecileRobertMichon) [inline pkg/acsengine/engine.go]: @jackfrancis is there a reason why this wasn't using the name var? is it okay that I changed it?

File: pkg/acsengine/engine.go [hunk 5]
```diff
@@ -1191,7 +1211,7 @@ func (t *TemplateGenerator) getTemplateFuncMap(cs *api.ContainerService) templat
 							val = dashboardAddon.Containers[dC].Image
 						}
 					} else {
-						val = cloudSpecConfig.KubernetesSpecConfig.KubernetesImageBase + KubeConfigs[k8sVersion]["dashboard"]
+						val = cloudSpecConfig.KubernetesSpecConfig.KubernetesImageBase + KubeConfigs[k8sVersion][DefaultDashboardAddonName]
```
Comment (CecileRobertMichon) [inline pkg/acsengine/engine.go]: @jackfrancis same here

File: pkg/acsengine/k8s_versions.go [hunk 6]
```diff
@@ -9,408 +9,423 @@ import (
...
-		"backoffretries":       strconv.Itoa(DefaultKubernetesCloudProviderBackoffRetries),
-		"backoffjitter":        strconv.FormatFloat(DefaultKubernetesCloudProviderBackoffJitter, 'f', -1, 64),
-		"backoffduration":      strconv.Itoa(DefaultKubernetesCloudProviderBackoffDuration),
-		"backoffexponent":      strconv.FormatFloat(DefaultKubernetesCloudProviderBackoffExponent, 'f', -1, 64),
-		"ratelimitqps":         strconv.FormatFloat(DefaultKubernetesCloudProviderRateLimitQPS, 'f', -1, 64),
-		"ratelimitbucket":      strconv.Itoa(DefaultKubernetesCloudProviderRateLimitBucket),
-		"gchighthreshold":      strconv.Itoa(DefaultKubernetesGCHighThreshold),
-		"gclowthreshold":       strconv.Itoa(DefaultKubernetesGCLowThreshold),
+		"hyperkube":                 "hyperkube-amd64:v1.8.2",
+		DefaultDashboardAddonName:   "kubernetes-dashboard-amd64:v1.7.1",
```
Comment (CecileRobertMichon) [inline pkg/acsengine/k8s_versions.go]: @jackfrancis and here (3/3)

File: pkg/api/const.go [hunk 7]
```diff
@@ -86,4 +86,6 @@ const (
 	DefaultTillerAddonEnabled = true
 	// DefaultDashboardAddonEnabled determines the acs-engine provided default for enabling kubernetes-dashboard addon
 	DefaultDashboardAddonEnabled = true
+	// DefaultReschedulerAddonEnabled determines the acs-engine provided default for enabling kubernetes-rescheduler addon
+	DefaultReschedulerAddonEnabled = true
```
Comment (CecileRobertMichon) [inline pkg/api/const.go]: True or false here?

File: pkg/acsengine/defaults.go (hunk 1 above)
Comment (jackfrancis) [reply pkg/acsengine/defaults.go]: Let's ship this as false to be cautious, and then figure out how we want to bake the feature before turning default to true.

File: pkg/acsengine/const.go [hunk 8]
```diff
@@ -85,6 +85,10 @@ const (
 	DefaultOrchestratorName = "k8s"
 	// DefaultEtcdDiskSize specifies the default size for Kubernetes master etcd disk volumes in GB
 	DefaultEtcdDiskSize = "128"
+	// DefaultReschedulerImage defines the rescheduler deployment version on Kubernetes Clusters
+	DefaultReschedulerImage = "rescheduler:v0.3.1"
```
Comment (CecileRobertMichon) [inline pkg/acsengine/const.go]: Is the right version to use as default?

File: pkg/acsengine/defaults.go (hunk 2 above)
Comment (jackfrancis) [reply pkg/acsengine/defaults.go]: We should copy what's here:

https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/rescheduler/rescheduler.manifest

(limits can copy requests)

File: pkg/acsengine/const.go (hunk 8 above)
Comment (jackfrancis) [reply pkg/acsengine/const.go]: let's ship w/ latest

File: pkg/acsengine/engine.go (hunk 4 above)
Comment (jackfrancis) [reply pkg/acsengine/engine.go]: 👍 and thanks!

File: pkg/acsengine/engine.go (hunk 5 above)
Comment (jackfrancis) [reply pkg/acsengine/engine.go]: 👍 and thanks ditto

File: pkg/api/const.go (hunk 7 above)
Comment (jackfrancis) [reply pkg/api/const.go]: This should be false to match.

File: pkg/acsengine/defaults.go (hunk 1 above)
Comment (jackfrancis) [reply pkg/acsengine/defaults.go]: Also, you could pay back some of my tech debt by changing these boolean literals inside `pointerToBool()` to `api.DefaultTillerAddonEnabled` and `api.DefaultDashboardAddonEnabled` :)

File: pkg/acsengine/const.go (hunk 8 above)
Comment (CecileRobertMichon) [reply pkg/acsengine/const.go]: 0.3.1 is the latest tag [here](https://console.cloud.google.com/gcr/images/google-containers/GLOBAL/rescheduler)

File: parts/kubernetesmastercustomdata.yml [hunk 9]
```diff
@@ -284,6 +293,13 @@ write_files:
     sed -i "s|<kubernetesTillerCPULimit>|{{WrapAsVariable "kubernetesTillerCPULimit"}}|g" "/etc/kubernetes/addons/kube-tiller-deployment.yaml"
     sed -i "s|<kubernetesTillerMemoryLimit>|{{WrapAsVariable "kubernetesTillerMemoryLimit"}}|g" "/etc/kubernetes/addons/kube-tiller-deployment.yaml"
 {{end}}
+{{if .OrchestratorProfile.KubernetesConfig.IsReschedulerEnabled}}
+    sed -i "s|<kubernetesReschedulerSpec>|{{WrapAsVariable "kubernetesReschedulerSpec"}}|g" "/etc/kubernetes/addons/kube-rescheduler-deployment.yaml"
+    sed -i "s|<kubernetesReschedulerCPURequests>|{{WrapAsVariable "kubernetesReschedulerCPURequests"}}|g" "/etc/kubernetes/addons/kube-rescheduler-deployment.yaml"
+    sed -i "s|<kubernetesReschedulerMemoryRequests>|{{WrapAsVariable "kubernetesReschedulerMemoryRequests"}}|g" "/etc/kubernetes/addons/kube-rescheduler-deployment.yaml"
+    sed -i "s|<kubernetesReschedulerCPULimit>|{{WrapAsVariable "kubernetesReschedulerCPULimit"}}|g" "/etc/kubernetes/addons/kube-rescheduler-deployment.yaml"
+    sed -i "s|<kubernetesReschedulerMemoryLimit>|{{WrapAsVariable "kubernetesReschedulerMemoryLimit"}}|g" "/etc/kubernetes/addons/kube-rescheduler-deployment.yaml"
+    {{end}}
```
Comment (jackfrancis) [inline parts/kubernetesmastercustomdata.yml]: for clarity let's unindent this `{{end}}` to match the `{{if``

File: pkg/acsengine/defaults.go [hunk 10]
```diff
@@ -334,6 +355,10 @@ func setOrchestratorDefaults(cs *api.ContainerService) {
 		if a.OrchestratorProfile.KubernetesConfig.Addons[d].IsEnabled(api.DefaultDashboardAddonEnabled) {
 			a.OrchestratorProfile.KubernetesConfig.Addons[d] = assignDefaultAddonVals(a.OrchestratorProfile.KubernetesConfig.Addons[d], DefaultDashboardAddonsConfig)
 		}
+		r := getAddonsIndexByName(a.OrchestratorProfile.KubernetesConfig.Addons, DefaultReschedulerAddonName)
+		if a.OrchestratorProfile.KubernetesConfig.Addons[r].IsEnabled(api.DefaultReschedulerAddonEnabled) {
+			a.OrchestratorProfile.KubernetesConfig.Addons[r] = assignDefaultAddonVals(a.OrchestratorProfile.KubernetesConfig.Addons[r], DefaultDashboardAddonsConfig)
```
Comment (jackfrancis) [inline pkg/acsengine/defaults.go]: The 2nd arg to `assignDefaultAddonVals()` here should be `DefaultReschedulerAddonsConfig`. (I know, copy/paste + tired eye sockets...)

File: pkg/acsengine/engine.go [hunk 11]
```diff
@@ -547,6 +548,7 @@ func getParameters(cs *api.ContainerService, isClassicMode bool, generatorCode s
 		addValue(parametersMap, "kubernetesDNSMasqSpec", cloudSpecConfig.KubernetesSpecConfig.KubernetesImageBase+KubeConfigs[k8sVersion]["dnsmasq"])
 		addValue(parametersMap, "kubernetesExecHealthzSpec", cloudSpecConfig.KubernetesSpecConfig.KubernetesImageBase+KubeConfigs[k8sVersion]["exechealthz"])
 		addValue(parametersMap, "kubernetesHeapsterSpec", cloudSpecConfig.KubernetesSpecConfig.KubernetesImageBase+KubeConfigs[k8sVersion]["heapster"])
+		addValue(parametersMap, "kubernetesReschedulerSpec", cloudSpecConfig.KubernetesSpecConfig.KubernetesImageBase+KubeConfigs[k8sVersion]["rescheduler"])
```
Comment (jackfrancis) [inline pkg/acsengine/engine.go]: use `DefaultReschedulerAddonName` as map key here

File: pkg/api/types.go [hunk 12]
```diff
@@ -599,3 +599,14 @@ func (k *KubernetesConfig) IsDashboardEnabled() bool {
 	}
 	return dashboardAddon.IsEnabled(DefaultDashboardAddonEnabled)
 }
+
+// IsReschedulerEnabled checks if the rescheduler addon is enabled
+func (k *KubernetesConfig) IsReschedulerEnabled() bool {
+	var reschedulerAddon KubernetesAddon
+	for i := range k.Addons {
+		if k.Addons[i].Name == "rescheduler" {
```
Comment (jackfrancis) [inline pkg/api/types.go]: use `DefaultReschedulerAddonName` here

File: pkg/api/types_test.go [hunk 13]
```diff
@@ -144,6 +144,36 @@ func TestIsDashboardEnabled(t *testing.T) {
...
+	c := KubernetesConfig{
+		Addons: []KubernetesAddon{
+			getMockAddon("addon"),
+		},
+	}
+	e := c.IsReschedulerEnabled()
+	if e != DefaultReschedulerAddonEnabled {
+		t.Fatalf("KubernetesConfig.IsReschedulerEnabled() should return %t when no rescheduler addon has been specified, instead returned %t", DefaultReschedulerAddonEnabled, e)
+	}
+	c.Addons = append(c.Addons, getMockAddon("rescheduler"))
```
Comment (jackfrancis) [inline pkg/api/types_test.go]: use `DefaultReschedulerAddonName` here

File: pkg/api/types_test.go [hunk 14]
```diff
@@ -144,6 +144,36 @@ func TestIsDashboardEnabled(t *testing.T) {
...
+	e = c.IsReschedulerEnabled()
+	if e != false {
+		t.Fatalf("KubernetesConfig.IsReschedulerEnabled() should return true when a custom rescheduler addon has been specified, instead returned %t", e)
+	}
+	b := true
+	c = KubernetesConfig{
+		Addons: []KubernetesAddon{
+			{
+				Name:    "rescheduler",
```
Comment (jackfrancis) [inline pkg/api/types_test.go]: use `DefaultReschedulerAddonName` here

File: pkg/acsengine/defaults.go (hunk 10 above)
Comment (CecileRobertMichon) [reply pkg/acsengine/defaults.go]: Yes thanks for catching that

File: parts/kubernetesmasteraddons-kube-rescheduler-deployment.yaml [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+      annotations:
+        scheduler.alpha.kubernetes.io/critical-pod: ''
+    spec:
+      containers:
+      - image: <kubernetesReschedulerSpec>
+        name: rescheduler
+        resources:
+          requests:
+            cpu: <kubernetesReschedulerCPURequests>
+            memory: <kubernetesReschedulerCPULimit>
```
Comment (jackfrancis) [inline parts/kubernetesmasteraddons-kube-rescheduler-deployment.yaml]: This needs to be `<kubernetesReschedulerMemoryRequests>`

File: docs/clusterdefinition.md [hunk 16]
```diff
@@ -49,6 +49,7 @@ Here are the valid values for the orchestrator types:
 |Name of addon|Enabled by default?|How many containers|Description|
 |tiller|true|Delivers the Helm server-side component: tiller. See https://github.com/kubernetes/helm for more info.|
 |kubernetes-dashboard|true|1|Delivers the kubernetes dashboard component. See https://github.com/kubernetes/dashboard for more info.|
+|rescheduler|false|Delivers the kubernetes rescheduler component.|
```
Comment (CecileRobertMichon) [inline docs/clusterdefinition.md]: I couldn't find the kubernetes repo for rescheduler

File: parts/kubernetesmasteraddons-kube-rescheduler-deployment.yaml (hunk 15 above)
Comment (CecileRobertMichon) [reply parts/kubernetesmasteraddons-kube-rescheduler-deployment.yaml]: oops yes thank you!

**Pull request merged**